File: 2.bfs_dfs/cpp/src/main.cpp
#include <iostream>
#include "node.hpp"

int main(int argc, char* argv[]) {
    FILE* input_file = std::fopen(argv[1], "r");

    int N; // number of nodes
    fscanf(input_file, "%d", &N);

    // dynamically allocate nodes
    Node** nodes = new Node*[N];
    for(int i=0;i<N;i++) {
        nodes[i] = new Node(i);
    }

    // initialize adjacent nodes
    int n_a, n_b;
    while(fscanf(input_file, "%d", &n_a)!=EOF) {
        fscanf(input_file, "%d", &n_b);
        nodes[n_a]->adj.push_back(nodes[n_b]);
        nodes[n_b]->adj.push_back(nodes[n_a]);
    }

    // bfs
    std::vector<int> bfs_trav_order;
    search::bfs(N, nodes, 0, bfs_trav_order);
    printf("bfs order: ");
    for(int idx: bfs_trav_order) {
        printf("%d ", idx);
    }
    printf("\n");

    // dfs
    std::vector<int> dfs_trav_order;
    search::dfs(N, nodes, 0, dfs_trav_order);
    printf("dfs order: ");
    for(int idx: dfs_trav_order) {
        printf("%d ", idx);
    }
    printf("\n");

    // free nodes
    for(int i=0;i<N;i++) {
        delete nodes[i];
    }
    delete[] nodes;

    return 0;
}

File: 2.bfs_dfs/python/main.py
import argparse
from bfs_dfs import bfs, dfs, Node

parser = argparse.ArgumentParser()
parser.add_argument("--input_file")
args = parser.parse_args()


if __name__ == "__main__":
    with open(args.input_file, "r") as f:
        N = int(f.readline())

        nodes = []
        for i in range(N):
            nodes.append(Node(i))

        while True:
            line = f.readline()
            if not line:
                break
            else:
                n_a, n_b = tuple(map(lambda x: int(x), line.split(" ")))
                nodes[n_a].adj.append(nodes[n_b])
                nodes[n_b].adj.append(nodes[n_a])

        bfs_order = []
        bfs(0, nodes, bfs_order)
        print("BFS order: {}".format(bfs_order)) 

        dfs_order = []
        dfs(0, nodes, dfs_order)
        print("DFS order: {}".format(dfs_order)) 